Upgrade TypeScript plugin, change 'ban-ts-comment' rule

mattpetrie opened the pull request:
# Changes
* Upgrade `@typescript-eslint/eslint-plugin` to the latest version, 4.6.1. (This was motivated by the rule change below, but seeing as we're two major versions behind, it seems like a good upgrade anyway.) (3dde9740c0c4436c6ece554dae3a3178fea53744)
* Change the [`@typescript/eslint/ban-ts-comment`](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/ban-ts-comment.md#allow-with-description) rule to have the 'allow-with-description' option for `@ts-ignore`. This is so we can stop doing [this](https://github.com/goodeggs/marketplace-mobile-app/search?q=%27eslint-disable-next-line+%40typescript-eslint%2Fban-ts-comment%27&unscoped_q=%27eslint-disable-next-line+%40typescript-eslint%2Fban-ts-comment%27) and [this](https://github.com/goodeggs/garbanzo/search?q=%27eslint-disable-next-line+%40typescript-eslint%2Fban-ts-comment%27&unscoped_q=%27eslint-disable-next-line+%40typescript-eslint%2Fban-ts-comment%27) and instead just leave a comment on the `@ts-ignore` as to why the line is being ignored. (https://github.com/goodeggs/eslint-plugin-goodeggs/pull/462/commits/fbe3d6df2075f25e339b938e3d1e49dd8320f394)
* Add configuration for the new [`@typescript-eslint/naming-convention`](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/naming-convention.md) rule that aligns with our [naming best practices](https://github.com/goodeggs/best-practices/blob/master/javascript/naming.md#acronym-naming). (a4a73357248e4847bf3f395e81ad99f4bd502a86)
* Add configuration for the [`@typescript-eslint/strict-boolean-expressions`](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/strict-boolean-expressions.md) rule. This should supersede https://github.com/goodeggs/eslint-plugin-goodeggs/pull/441 (the commits are cherry-picked here as that configuration appears to still be what we want). (https://github.com/goodeggs/eslint-plugin-goodeggs/pull/462/commits/f5c89716341919651b21134e16e7f322c96735ab and 0bfba76)
* Make TypeScript >=4.0.5 an optional peer dependency.
* Updated changelog with these changes. (3dde9740c0c4436c6ece554dae3a3178fea53744)

Companion GEK PR: https://github.com/goodeggs/goodeggs-toolkit/pull/572

Fixes: https://github.com/goodeggs/standards-and-best-practices/issues/232.

serhalp commented on src/config/typescript.js:
```diff
@@ -29,7 +29,7 @@ export default {
         'no-useless-constructor': 'off',
 
-        '@typescript-eslint/ban-ts-comment': 'error',
+        '@typescript-eslint/ban-ts-comment': ['error', {'ts-ignore': 'allow-with-description'}],
```
I think that makes sense to me. The goal is to make it especially hard to casually `@ts-ignore` without being sure you really want that and without documenting why and/or when we can remove it. Requiring a comment seems reasonable to me.

serhalp commented:
Looking at this more carefully after your comments in https://github.com/goodeggs/goodeggs-toolkit/pull/572, I see now there are likely various breaking changes here. We should review these: https://github.com/typescript-eslint/typescript-eslint/releases/tag/v3.0.0. Anything that's worth noting for consuming repos should be documented in the eslint-plugin-goodeggs changelog. Anything that requires a config change in eslint-plugin-goodeggs should be updated in this PR. There's also some trickier cases like the breaking changes to `strict-boolean-expressions` which affect this open PR: https://github.com/goodeggs/eslint-plugin-goodeggs/pull/441.

We should probably roll this into the queued up PRs with breaking changes that we want to package up into a single release?

mattpetrie commented:
@serhalp I did a review of the new and changed rules in v3.9 and added changes here for 2 more rules, noted in the PR description above. I also updated the changelog. Might be worthy of a re-review.

mattpetrie commented on src/config/typescript.js:
```diff
@@ -28,14 +27,33 @@ export default {
+          },
+          // Only capitalize the first letter of each section of a class or interface name.
+          // e.g. Bad: "HTTPAPIContractTestHelper", Good: "HttpApiContractTestHelper".
+          // This is enforced by disallowing any name with 2 capital letters in a row.
```
Is this too draconian? The more I've thought about it, I worry there may be too many legitimate exceptions to this rule. For example, I know there are many places in MMA where we have used `graphQL`, which I personally feel reads better than `graphql` or `graphQl`, but I don't feel strongly about it. One way this could be made more forgiving is to make 3 consecutive capitals the threshold instead of two, which would still ban `HTTP` and `API`, for example, but would allow `graphQL`. Thoughts?

mattpetrie commented:
@serhalp I have updated this to PR to use the latest versions of `@typescript-eslint` and to have a peer dependency on Typescript 4. I have also made the same changes to:  https://github.com/goodeggs/goodeggs-toolkit/pull/572.

serhalp reviewed (approved):
🚀

serhalp commented on src/config/typescript.js:
```diff
@@ -32,10 +31,33 @@ export default {
         // Enabled by @typescript-eslint/recommended, conflicts with prettier
         '@typescript-eslint/indent': 'off',
+
+        '@typescript-eslint/naming-convention': [
```
@mattpetrie hah, you had actually added this but then we removed it before merging

mattpetrie replied:
🤦 Oh yeah. I had decided the custom rules I defined were too strict, but that probably didn't warrant removing the rule altogether!
